### PR TITLE
Smart Charging - SetChargingProfile object and data types

### PR DESCRIPTION
**Pull request by LiorBenAri**

### CODE REVIEW

**Comment by YuriyDurov** [comment]
Hello @LiorBenAri! This PR looks great, I just have a small issue with a part of it:

OCPI.Net uses contract validators for validating the presense of properties instead of marking properties as required in the contract model. Normally, we define all contract properties as nullable and then use contract validators to check whether a property was null or not. This approach allows us to better distinguish between null vs default values (like 0 for numbers) in incoming data, and also allows us to better form validation error messages in case if some properties did not satisfy the protocol requirements. It also allows us to better validate things like arrays (at least one item present, etc.) and define other specific validation rules.

Can we have this PR updated so that it would not define `required` properties in contract models but would use validators to check for null values instead?

[Here](https://bitzart.github.io/OCPI.Net/4.contracts.html) is some OCPI.Net documentation on this topic

And Validators code is located under `src/OCPI.Net.Validation/Validators`

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    public override string ToString()
+    {
+        var toStringBuilder = new StringBuilder();
+        toStringBuilder.Append($"Start date time: {StartDateTime}, ");
+        toStringBuilder.AppendLine($"Duration: {Duration},");
+        toStringBuilder.Append($"Charging rate unit: {ChargingRateUnit}, ");
+        toStringBuilder.AppendLine($"Minimum charging rate: {MinChargingRate},");
+        var chargingProfilePeriodToString = ChargingProfilePeriodsToString();
+        return toStringBuilder.Append(chargingProfilePeriodToString).ToString();
```
Comment (YuriyDurov) [inline src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs:58]: I would suggest simlpifying this a bit:
```csharp
toStringBuilder.Append(ChargingProfilePeriodsToString());
return toStringBuilder.ToString();
```

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private string ChargingProfilePeriodsToString()
+    {
```
Comment (YuriyDurov) [inline src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs:62]: How about something like this?
```csharp
if (ChargingProfilePeriod is null || !ChargingProfilePeriod.Any()) return string.Empty;
```
This way you wouldn't have to check if it is null later and also wouldn't have to create the StringBuilder if it is null or no periods are present

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiSetChargingProfileCommand.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Text.Json.Serialization;
+
+namespace OCPI.Contracts.ChargingProfiles;
+public class OcpiSetChargingProfileCommand
```
Comment (YuriyDurov) [inline src/OCPI.Net.Contracts/ChargingProfiles/OcpiSetChargingProfileCommand.cs:4]: I understand where the idea of calling it a `command` comes from but I am worried about it potentially creating confusion between the commands featured in the Commands Module and this. How about we name
 it `OcpiSetChargingProfileRequest` instead?

**Comment by LiorBenAri** [comment]
> Hello @LiorBenAri! This PR looks great, I just have a small issue with a part of it:
> 
> OCPI.Net uses contract validators for validating the presense of properties instead of marking properties as required in the contract model. Normally, we define all contract properties as nullable and then use contract validators to check whether a property was null or not. This approach allows us to better distinguish between null vs default values (like 0 for numbers) in incoming data, and also allows us to better form validation error messages in case if some properties did not satisfy the protocol requirements. It also allows us to better validate things like arrays (at least one item present, etc.) and define other specific validation rules.
> 
> Can we have this PR updated so that it would not define `required` properties in contract models but would use validators to check for null values instead?
> 
> [Here](https://bitzart.github.io/OCPI.Net/4.contracts.html) is some OCPI.Net documentation on this topic
> 
> And Validators code is located under `src/OCPI.Net.Validation/Validators`

Hi Yuri,

Will edit as you requested.

Lior

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs (hunk 1 above)
Comment (LiorBenAri) [reply src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs]: done

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiSetChargingProfileCommand.cs (hunk 3 above)
Comment (LiorBenAri) [reply src/OCPI.Net.Contracts/ChargingProfiles/OcpiSetChargingProfileCommand.cs]: done

File: src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs (hunk 2 above)
Comment (LiorBenAri) [reply src/OCPI.Net.Contracts/ChargingProfiles/OcpiChargingProfile.cs:66]: done

**Comment by LiorBenAri** [comment]
> > Hello @LiorBenAri! This PR looks great, I just have a small issue with a part of it:
> > OCPI.Net uses contract validators for validating the presense of properties instead of marking properties as required in the contract model. Normally, we define all contract properties as nullable and then use contract validators to check whether a property was null or not. This approach allows us to better distinguish between null vs default values (like 0 for numbers) in incoming data, and also allows us to better form validation error messages in case if some properties did not satisfy the protocol requirements. It also allows us to better validate things like arrays (at least one item present, etc.) and define other specific validation rules.
> > Can we have this PR updated so that it would not define `required` properties in contract models but would use validators to check for null values instead?
> > [Here](https://bitzart.github.io/OCPI.Net/4.contracts.html) is some OCPI.Net documentation on this topic
> > And Validators code is located under `src/OCPI.Net.Validation/Validators`
> 
> Hi Yuri,
> 
> Will edit as you requested.
> 
> Lior

I see that there isn't a validation logic for all of the library objects (for example Session\CDRs objects).
Do you consider adding validation logic for the ChargingProfile objects as mandatory?

**Comment by LiorBenAri** [comment]
Thank you Yuri.
I like you to know that the package is great and it is super helpful for us for our OCPI development!

**Comment by LiorBenAri** [comment]
Thank you :)

**Pull request merged**